url8417b1b213_Ratchette_Rummikub_tgz-pJ8-rummikub_RummikubJ8/src/rummikub/GameInfo.java

```diff
@@ -0,0 +1,435 @@
+	 */
+	public void addTile(int player){
+		tilesInHand[player]++;
+	}
+
+	public void setHand(int player, int size){
+		tilesInHand[player] = size;
+	}
+	
+	public void addMelds(ArrayList<Meld> newMelds){
+		board.addAll(newMelds);
+	} 	
+	
+	public void setMelds(ArrayList<Meld> melds){
+		ArrayList<Tile> oldTiles, newTiles;
+		
+		oldTiles = new ArrayList<Tile>();
+		for(Meld meld : board)
+			oldTiles.addAll(meld.getTiles());
+		
+		newTiles = new ArrayList<Tile>();
+		for(Meld meld: melds)
+			newTiles.addAll(meld.getTiles());
+
+		for(Tile tile : oldTiles)
+			newTiles.remove(tile);
+		
+		board = new ArrayList<Meld>(melds);
+	}
+	
+	/**
+	 * Determine if the game has been won yet
+	 * @return true if any player has no cards remaining in their hand
+	 */
+	public boolean isGameOver(){
+		for(int i=0; i<tilesInHand.length; i++)
+			if(tilesInHand[i] < 1)
+				return true;
+		
+		return false;
+	}
+	
+	public int getWinner(){
+		for(int i=0; i<tilesInHand.length; i++)
+			if(tilesInHand[i] < 1)
+				return i;
+		
+		return DISCONNECT;
+	}
+	
+	public ArrayList<Meld> getBoard(){
+		return new ArrayList<Meld>(board);
+	}
+
+	// **********************************************************
+    //						Move creation							
+    // **********************************************************
+
+	// combine hand with deck and find all valid moves
+	public ArrayList<Meld> getBruteForceMove(Hand hand, int playerNum) throws Exception{
+		ArrayList<Meld> melds;
+		ArrayList<Tile> originalTiles, usedTiles;
+		Hand allTiles;
+		
+		// add all tiles on the board and in your hand to one set
+		originalTiles = new ArrayList<Tile>();
+		originalTiles.addAll(hand.getTiles());
+		for(int i=0; i<board.size(); i++)
+			originalTiles.addAll(board.get(i).getTiles());
+		
+		// generate all possible runs + groups
+		allTiles = new Hand(originalTiles);
+		melds = allTiles.getGroups();
+		melds.addAll(allTiles.getRuns());
+		
+		// get the mutually exclusive subset with the largest score
+		melds = allTiles.findLargestSubset(new ArrayList<Tile>(allTiles.getTiles()), new ArrayList<Tile>(), melds);
+
+		// Create a list of tiles from the original board
+		originalTiles = new ArrayList<Tile>();
+		for(int i=0; i<board.size(); i++)
+			originalTiles.addAll(board.get(i).getTiles());
+		
+		// make a list of all tiles used in the new melds
+		usedTiles = new ArrayList<Tile>();
+		for(Set set : melds)
+			usedTiles.addAll(set.getTiles());
+		
+		// remove tiles from the original board from the tiles that you used
+		for(Tile tile : originalTiles){
+			if(usedTiles.contains(tile))
+				usedTiles.remove(tile);
+		}
+		
+		// no tiles were moved from your hand to the board
+		if(usedTiles.size() == 0)
+			return null;
+		
+		// remove the tiles that you played from your hand
+		for(Tile tile : usedTiles){
+			hand.removeTile(tile);
+			tilesInHand[playerNum]--;
+		}
+		
+		return melds;
+	}
+	
+	public ArrayList<Meld> getAdjacentPlay(Hand hand, int playerNum) throws Exception{
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.